src/pudl_archiver/archivers/classes.py

```diff
@@ -241,11 +246,27 @@ async def download_all_resources(self) -> dict[str, ResourceInfo]:
         # Get all awaitables from get_resources
         resources = [resource async for resource in self.get_resources()]
 
-        # Download resources concurrently and prepare metadata
-        resource_dict = {}
-        for resource_coroutine in asyncio.as_completed(resources):
-            resource_info = await resource_coroutine
-            self.logger.info(f"Downloaded {resource_info.local_path}.")
-            resource_dict[str(resource_info.local_path.name)] = resource_info
+        # Split resources into chunks to limit concurrency
+        chunksize = self.concurrency_limit if self.concurrency_limit else len(resources)
+        resource_chunks = [
+            resources[i * chunksize : (i + 1) * chunksize]
+            for i in range(math.ceil(len(resources) / chunksize))
+        ]
 
-        return resource_dict
+        if self.concurrency_limit:
+            self.logger.info("Downloading resources in chunks to limit concurrency")
+            self.logger.info(f"Resource chunks: {len(resource_chunks)}")
+            self.logger.info(f"Resources per chunk: {chunksize}")
+
+        # Download resources concurrently and prepare metadata
+        for resource_chunk in resource_chunks:
+            for resource_coroutine in asyncio.as_completed(resource_chunk):
+                resource_info = await resource_coroutine
+                self.logger.info(f"Downloaded {resource_info.local_path}.")
+                yield str(resource_info.local_path.name), resource_info
+
+            # If requested, create a new temporary directory per resource chunk
+            if self.directory_per_resource_chunk:
+                tmp_dir = tempfile.TemporaryDirectory()
```

Comment:
The goal of this is to limit disc usage for large datasets to avoid overwhelming the github runner disc limits, so it's actually intentionally blowing away the directory after every chunk is finished being processed (`temp_dir` gets overwritten with every iteration and the GC will clean up the old directory). This is all configurable though, so datasets can continue to use the old directory structure if they don't need this.

Reply:
Sweet, thanks for the explanation!